### ISSUE TITLE
Why masking is done after xoring round_dout and din ?

### ISSUE THREAD

**Issue by atkarim** [pmassolino/hw-subterranean#1]
Hi,

In the spec of Subterranean 2.0, the masking is made over round_dout and then the result is xored with din.
![image](https://user-images.githubusercontent.com/45005878/131848174-5c9ea68c-cd7a-4c27-ba44-b86a3f54e5e7.png)

Like in this old [verilog source](https://cs.ru.nl/~joan/Subterranean/subterranean_verilog.zip), we have :

`temp_din_1_xor_dout = (round_1_dout & round_1_dout_mask) ^ temp_din_1`

But in your implementation, we have the xoring before :

`temp_din_1_xor_dout = (round_1_dout ^ temp_din_1) & round_1_dout_mask`

I want to know, if there is a reason for this change, or if it's a paranthesis error.

**Comment by atkarim** [comment]
I agree that the two implementation will give the same result, if temp_din_1 has the correct din_size.

**Comment by pmassolino** [comment]
Hi,

You have already answered the question. Both approaches are correct and give the same output if the input has zeros in the non-populated position, which is an okay assumption. However, by using the new order of operands the new implementation is correct even if the input doesn't have these zeroes. 

I honestly don't remember exactly why I changed the order. The most probable reason is because of a possible optimization in the higher level design. If the inner design doesn't need the inputs to have the 0's in that position, the higher level design does not need to add. Another possibility was that there was a bug that I found out really late when integrating the higher level design and this part. If there was such a bug, I most likely warned on NIST LWC mailing list. Anyway, the design here on Github is correct and most up to date, if you want to use one, then take the version here.

About the drawing, there is a new drawing that reflects this change. The drawing is not here, but I can easily add if required.
